### DISCUSSION TITLE
Is there attribute that shows date when photo was added to Photos app?

### DISCUSSION THREAD

**Discussion by ghost** [RhetTbull/osxphotos#438] [Q&A] [answered]
Hi there,

 my use case is the following:
I want to create every day at 12:01am a backup of all photos that were added the day before.

For that I use '```from_date```` to find photos from the day before.
The flaw: I might add photos taken by another camera then my iPhone sometime days or weeks later.
The problem is that from_date looks at the photo creation date and not that date the photo was added to Photos app.

I tried to find an appropriate in the doc but it seems there is no such attribute available.
So question is:
1. In case I have overlooked it: what is the name?
2. In case it does not exist, is it possible to add it?
3. What would be a workaround (I was thinking into the direction to compare the filenames, etc. but that would mean I had to always check against the whole library of +44k photos - not very efficient)?

Thanks a lot,
Kevin

**Comment by RhetTbull** [comment, marked as answer]
@hafur there's not currently a way to access the date added.  It won't be hard to to do this so I'll add it to the to-do list. The relevant data appears to be in `ZGENERICASSET.ZADDEDDATE` but I'll have to do some testing to confirm.

However, for your use case, `osxphotos export --update` may get you close.  `--update` exports only the photos not previously exported, thus if you run this every day, it would only export those photos added the previous day. 

*Note* `--update` stores state information about the backup / export in a file called `.osxphotos_export.db` in the export directory.  So this approach only works if you're nightly backups are all in the same directory.

**Reply by ghost** [reply]
great! thanks a lot! I'll be patient and wait for the implementation. I do some other things as part of the export in Python and would not want to rely on the command line right now.
thanks again!

**Comment by RhetTbull** [comment]
@hafur In the latest version on Github I've added a PhotoInfo.date_added attribute.  It returns a timezone aware `datetime.datetime` object as do the other dates in PhotoInfo.

Caveats: 

- this only works for Photos 5+ (Catalina+) as I've not figured out where the date is stored in Photos 4 which has a different schema
- On Photos <= 4, returns `None`
- I've not yet added any tests or documentation 
- I've not yet pushed a new release (won't until tests/docs updated)

If you use this and have any feedback, let me know!  I'll get this into the next release.

**Reply by ghost** [reply]
man thanks a lot! that was quick!

now the bad news: I am running my workflow on a 2011 Mac Mini which only support Photos 3 :-(
the good news: I think you just convinced me to repurpose my old 2014 Mac Book Po to become my new home server and invest in a new M1 Mac Book :-) 
(it also helps with my workflow as I had done a lot of extra work to get to the ExiF infos as part of my backup process).

Thanks again!!!!

**Reply by RhetTbull** [reply]
I'm delighted to see another developer using osxphotos.  My original intent for writing this was to enable other developers to build their own workflows.  But -- I've found almost everyone just wants the command line so that's where I've focused my efforts.  I'm curious if the command line tool was missing something you needed for your workflow or if you just wanted to roll your own?

**Reply by ghost** [reply]
i probably could use the command line tool as well - but in the end I like tinkering around with Python ;-)
My biggest challenge is that my 24/7 server is just this old MacMini running Photos 3. 
My workload looks like this:
1. Daily backups of everything that was created the day before and store it in folders by month
2. Convert any HEIC to JPG - dont need no HEIC ;-) - I am using my Python scripts as a wrapper to invoke heic2jpg
3. Every photo should be stored in a folder named after the camera model that was used to take the photo, e.g. iPhone, Fuji, Ricoh, etc. - I using exif module and not exifTool (too complicated for me ;-)
4. Photos that do not contain Exif info go into a dedicated folder such as PNG files
5. movies go into yet another folder

Now that you are adding the 'date_added' attribute, I am definitely looking to upgrade my server. Which means I can also rely on the exif method in osxphotos - would make things even easier.

Now what would be cheery on the cake if osxphotos could also convert HEIC to JPG - but I guess that might be a license issue (?)

On my own little roadmap is to extract the complete Exif information and store it (as JSON) together with the image (as BLOB) in an Oracle Autonomous DB part of the -always free- offering. 
Once that is done I was thinking to see what could be done with the Exif info and some machine learning. Who knows maybe it can predict what photo I am going to take next :-)

**Reply by RhetTbull** [reply]
>but in the end I like tinkering around with Python ;-)

Me too!  

> Now what would be cheery on the cake if osxphotos could also convert HEIC to JPG - but I guess that might be a license issue (?)

osxphotos does this!  on the command line, you use the `--convert-to-jpeg` option.   For your own code, you can use osxphotos [ImageConverter](https://github.com/RhetTbull/osxphotos/blob/master/osxphotos/imageconverter.py) class (not documented but it's what osxphotos uses internally).  An example of how to use it here:

https://github.com/RhetTbull/osxphotos/blob/b4bc906b6a1c3444c5f5a5d9d908ab8c955c8f7e/osxphotos/fileutil.py#L192-L207

`ImageConverter` uses Apple's CoreGraphics framework to do the conversion so it can convert any format the Mac natively supports (which is almost all of them)!

`exiftool` is complicated but OSXPhotos also includes an ExifTool wrapper class (available via the [PhotoInfo.exiftool](https://github.com/RhetTbull/osxphotos/blob/b4bc906b6a1c3444c5f5a5d9d908ab8c955c8f7e/osxphotos/fileutil.py#L192-L207) attribute)

> On my own little roadmap is to extract the complete Exif information and store it (as JSON)

```python
photo.exiftool.json()
```
:-)   Of course, need to check to see that `photo.exiftool` isn't `None` which it will be if the file is missing from the Photos library.  All you need to use this for exiftool to be in your path.

**Reply by ghost** [reply]
oh wow! nice - gotta go now - need to refactor my code :-)
thanks for the hints. very cool!

**Reply by ghost** [reply]
Heya @RhetTbull quick question on the ImageConverter.
I tried it but I get this error:
```
Traceback (most recent call last):
  File "/Users/kevin/Documents/python_projects/photobackuptool/converter.py", line 14, in <module>
    main()
  File "/Users/kevin/Documents/python_projects/photobackuptool/converter.py", line 9, in main
    converter.write_jpeg('/Volumes/YOGI/tmp/tmp/20210403_IMG_0045.HEIC','/Volumes/YOGI/tmp/tmp/20210403_IMG_0045.jpg')
  File "/usr/local/lib/python3.9/site-packages/osxphotos/imageconverter.py", line 117, in write_jpeg
    ) = self.context.writeJPEGRepresentationOfImage_toURL_colorSpace_options_error_(
AttributeError: 'NoneType' object has no attribute 'writeJPEGRepresentationOfImage_toURL_colorSpace_options_error_'
```

Any suggestions?

This was the test I ran:
```
converter = osxphotos.fileutil.ImageConverter()
    converter.write_jpeg('/Volumes/YOGI/tmp/tmp/20210403_IMG_0045.HEIC','/Volumes/YOGI/tmp/tmp/20210403_IMG_0045.jpg')
```

Thanks a lot!

**Reply by RhetTbull** [reply]
@hafur  give this a try:

```pycon
>>> from osxphotos.imageconverter import ImageConverter
>>> converter = ImageConverter()
>>> converter.write_jpeg("/users/rhet/Downloads/test.heic","/users/rhet/Downloads/test.jpeg")
True
>>>
```

You can access the image converter through FileUtil class which is a special class with several utilities osxphotos uses to enable the --dry-run feature (there's a version of this class with no-op methods that don't actually do anything that can be dropped in when using --dry-run).  To use it from FileUtil, try the following--note the convert_to_jpeg method which calls ImageConverter is a class method so you don't need a FileUtil instance:

```pycon
>>> from osxphotos.fileutil import FileUtil
>>> FileUtil.convert_to_jpeg("/users/rhet/Downloads/test.heic","/users/rhet/Downloads/test.jpeg")
True
>>>
```

**Reply by ghost** [reply]
thanks a bunch!
I tried your suggestions on my Mac mini (running High Sierra hence Photos v3) and throws the same error.
However, on my Big Sur MacBook all variants (that is my code and your code) work fine. So it seems I have found yet one more reason to upgrade my Mac mini :-)

**Reply by RhetTbull** [reply]
Interesting.  It should work fine on High Sierra as far as a I know.  I just tried it on my 2013 iMac with Sierra and it worked. (Like you, my primary photos workflow is on an older Mac...hence once reason I try to maintain backwards compatibility all the way back to Sierra for everything I do in osxphotos).  However, I do focus new development on Catalina+ so won't spend much time trying to figure out why it didn't work in your case.  The API call is supported on all MacOS versions but it does require a GPU.  I looked up the tech specifications for 2010-2011 Mac Minis and they do have GPUs.

**Reply by RhetTbull** [reply]
By the way, I figured out how to get the "date added" in Photos 3 but it will take more work to implement than it did for Photos 5.  I'll get to it when I can.  But you should still upgrade your Mac :-)

**Reply by ghost** [reply]
thanks so much for that - really appreciated.
however, I have decided to upgrade. There are so many advantages. So for now I keep using my MacBook to create those daily backups (which are then transferred to the Mac mini's raid cluster).
BTW regarding the ImageConverter(): Could the issue be that I am running my MacMini headless? I think I remember reading ling time ago that the MacMini acts up when having no HDMI cable connected to it. I will try over the weekend if I can get a monitor connected and try it again.
Thanks again!

**Comment by ghost** [comment]
btw: the new attribute ```date_added``` works perfectly! Thanks lot!

**Comment by RhetTbull** [comment]
`date_added` now working for Photos >= 2 as of release [0.42.17](https://github.com/RhetTbull/osxphotos/releases/tag/v0.42.17)
